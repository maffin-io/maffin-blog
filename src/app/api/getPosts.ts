import { DateTime } from 'luxon';
import { Octokit } from 'octokit';

import markdownToHtml from '@/lib/markdownToHtml';

export type Post = {
  slug: string,
  date: DateTime,
  title: string,
  summary: string,
  reading_time: string,
  tags: string[],
  content: string,
  author: {
    name?: string,
    avatar?: string,
  },
};

export async function getPosts(): Promise<Post[]> {
  const octokit = new Octokit();
  const issues = (await octokit.rest.issues.listForRepo({
    owner: 'maffin-io',
    repo: 'maffin-blog',
    creator: 'argaen',
    headers: {
      accept: 'application/vnd.github+json',
    },
  })).data.filter(issue => !issue.pull_request && !issue.title.includes('[draft]'));

  return Promise.all(issues.map(async (issue) => {
    if (!issue.body) {
      throw new Error(`Missing body in issue ${issue.title}`);
    }

    const { content: htmlContent, metadata } = await markdownToHtml(issue.body);
    return {
      slug: metadata.slug,
      date: DateTime.fromISO(issue.created_at),
      title: issue.title,
      summary: metadata.summary,
      reading_time: metadata.reading_time,
      // @ts-ignore
      tags: issue.labels.map(label => label.name),
      content: htmlContent,
      author: {
        name: issue.user?.name || issue.user?.login,
        avatar: issue.user?.avatar_url,
      },
    };
  }));
}

export async function getPost(slug: string): Promise<null | Post> {
  const posts = await getPosts();
  return posts.find(post => post.slug === slug) || null;
}

export async function getPostsByTag(tag: string): Promise<Post[]> {
  const posts = await getPosts();
  return posts.filter(post => post.tags.includes(tag));
}
