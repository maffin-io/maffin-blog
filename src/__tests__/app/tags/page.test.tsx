import { render } from '@testing-library/react';

import TagsPage from '@/app/tags/page';
import * as postsApi from '@/app/api/getPosts';
import type { Post } from '@/app/api/getPosts';

jest.mock('@/app/api/getPosts', () => ({
  __esModule: true,
  ...jest.requireActual('@/app/api/getPosts'),
}));

/**
 * @param {function} Component
 * @param {*} props
 * @returns {Promise<()=>JSX.Element>}
 */
async function resolveComponent(Component: React.FunctionComponent, props: any) {
  const ComponentResolved = await Component(props);
  return () => ComponentResolved;
}

describe('HomePage', () => {
  it('renders as expected', async () => {
    jest.spyOn(postsApi, 'getPosts').mockResolvedValue([
      {
        tags: ['label1'],
      } as Post,
      {
        tags: ['label1', 'label2'],
      } as Post,
    ]);

    const Component = await resolveComponent(TagsPage, {});
    const { container } = render(<Component />);

    expect(container).toMatchSnapshot();
  });
});
