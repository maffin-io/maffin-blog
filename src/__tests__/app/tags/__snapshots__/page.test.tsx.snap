// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsPage renders as expected 1`] = `
<div>
  <div
    class="divide-y divide-gray-200"
  >
    <div
      class="space-y-2 pt-6 pb-8 md:space-y-5"
    >
      <h1
        class="text-3xl font-extrabold leading-9 tracking-tight text-gray-900 sm:text-4xl sm:leading-10 md:text-6xl md:leading-14"
      >
        Tags
      </h1>
    </div>
    <ul
      class="divide-y divide-gray-200 dark:divide-gray-700"
    >
      <li
        class="py-12"
      >
        <a
          href="/tags/label1"
        >
          <span
            class="mr-3 text-sm font-medium uppercase text-primary-500 hover:text-primary-600 dark:hover:text-primary-400"
          >
            #
            label1
          </span>
        </a>
        2
         
        posts
      </li>
      <li
        class="py-12"
      >
        <a
          href="/tags/label2"
        >
          <span
            class="mr-3 text-sm font-medium uppercase text-primary-500 hover:text-primary-600 dark:hover:text-primary-400"
          >
            #
            label2
          </span>
        </a>
        1
         
        posts
      </li>
    </ul>
  </div>
</div>
`;
